src/Microsoft.Azure.WebJobs.Host/Bindings/QueueBindingRules.cs

```diff
@@ -0,0 +1,217 @@
+        {
+            string queueName = attribute.QueueName;
+            if (nameResolver != null)
+            {
+                queueName = nameResolver.ResolveWholeString(queueName);
+            }
+            queueName = queueName.ToLowerInvariant(); // must be lowercase. coerce here to be nice.
+            return queueName;
+        }
+
+        // This is a static validation (so only %% are resolved; not {} ) 
+        // For runtime validation, the regular builder functions can do the resolution.
+        private static void ValidateQueueAttribute(QueueAttribute attribute, INameResolver nameResolver)
+        {
+            string queueName = NormalizeQueueName(attribute, nameResolver);
+            QueueClient.ValidateQueueName(queueName);
+        }
+
+        private IStorageQueueMessage ConvertReal2CloudQueueMessage(CloudQueueMessage arg)
+        {
+            return new StorageQueueMessage(arg);
+        }
+
+        private byte[] ConvertCloudQueueMessage2ByteArray(IStorageQueueMessage arg)
+        {
+            return arg.AsBytes;
+        }
+
+        private string ConvertCloudQueueMessage2String(IStorageQueueMessage arg)
+        {
+            return arg.AsString;
+        }
+
+        private IStorageQueueMessage ConvertByteArrayCloudQueueMessage(byte[] arg, QueueAttribute attrResolved)
+        {
+            var attr = (ResolvedQueueAttribute)attrResolved;
+            var msg = attr.GetQueue().CreateMessage(arg);
+            return msg;
+        }
+
+        private IStorageQueueMessage ConvertString2CloudQueueMessage(string arg, QueueAttribute attrResolved)
+        {
+            var attr = (ResolvedQueueAttribute)attrResolved;
+            var msg = attr.GetQueue().CreateMessage(arg);
+            return msg;
+        }
+
+        private async Task<CloudQueue> BuildRealClientFromQueueAttributeAsync(QueueAttribute attrResolved)
+        {
+            var queue = await this.BuildClientFromQueueAttributeAsync(attrResolved);
+            return queue.SdkObject;
+        }
+
+        private async Task<IStorageQueue> BuildClientFromQueueAttributeAsync(QueueAttribute attrResolved)
+        {
+            var attr = (ResolvedQueueAttribute)attrResolved;
+            var queue = attr.GetQueue();
+            await queue.CreateIfNotExistsAsync(CancellationToken.None);
+            return queue;
+        }
+
+        private IAsyncCollector<IStorageQueueMessage> BuildFromQueueAttribute(QueueAttribute attrResolved)
+        {
+            var attr = (ResolvedQueueAttribute)attrResolved;
+            var queue = attr.GetQueue();
+            return new QueueAsyncCollector(queue, _messageEnqueuedWatcherGetter.Value);
+        }
+
+        // Queue attributes are paired with a separate [StorageAccount]. 
+        // Consolidate the information from both attributes into a single attribute.
+        internal sealed class ResolvedQueueAttribute : QueueAttribute
```

Comment:
We're going to need to find a better pattern for this. While you were able to work around this specific instance with some hacks (I think you'll agree they're hacks :)) other instances like this are going to arise. I'm wondering if the way out isn't to introduce a context object that gets flowed around through your new pipeline, along with the attribute. I think we're going to need something like that - otherwise people are going to just start hanging all kinds of crap of their attributes, abusing the .NET attribute model and treating the attributes as context bags.